core/src/main/scala/org/apache/spark/status/LiveEntity.scala

```diff
@@ -141,96 +145,95 @@ private class LiveTask(
-      metrics.shuffleReadMetrics.recordsRead - old.shuffleReadMetrics.recordsRead)
-
-    val inputDelta = new v1.InputMetrics(
-      metrics.inputMetrics.bytesRead - old.inputMetrics.bytesRead,
-      metrics.inputMetrics.recordsRead - old.inputMetrics.recordsRead)
-
-    val outputDelta = new v1.OutputMetrics(
-      metrics.outputMetrics.bytesWritten - old.outputMetrics.bytesWritten,
-      metrics.outputMetrics.recordsWritten - old.outputMetrics.recordsWritten)
-
-    new v1.TaskMetrics(
-      0L, 0L,
-      metrics.executorRunTime - old.executorRunTime,
-      metrics.executorCpuTime - old.executorCpuTime,
-      0L, 0L, 0L,
-      metrics.memoryBytesSpilled - old.memoryBytesSpilled,
-      metrics.diskBytesSpilled - old.diskBytesSpilled,
-      0L,
-      inputDelta,
-      outputDelta,
-      shuffleReadDelta,
-      shuffleWriteDelta)
-  }
-
-  override protected def doUpdate(): Any = {
+  private def buildUpdate(): TaskDataWrapper = {
     val duration = if (info.finished) {
       info.duration
     } else {
       info.timeRunning(lastUpdateTime.getOrElse(System.currentTimeMillis()))
     }
 
-    val task = new v1.TaskData(
+    new TaskDataWrapper(
       info.taskId,
       info.index,
       info.attemptNumber,
-      new Date(info.launchTime),
-      if (info.gettingResult) Some(new Date(info.gettingResultTime)) else None,
-      Some(duration),
-      info.executorId,
-      info.host,
-      info.status,
-      info.taskLocality.toString(),
+      info.launchTime,
+      if (info.gettingResult) info.gettingResultTime else -1L,
+      duration,
+      weakIntern(info.executorId),
+      weakIntern(info.host),
+      weakIntern(info.status),
+      weakIntern(info.taskLocality.toString()),
       info.speculative,
       newAccumulatorInfos(info.accumulables),
       errorMessage,
-      Option(recordedMetrics))
-    new TaskDataWrapper(task, stageId, stageAttemptId)
+
+      metrics.executorDeserializeTime,
+      metrics.executorDeserializeCpuTime,
+      metrics.executorRunTime,
+      metrics.executorCpuTime,
+      metrics.resultSize,
+      metrics.jvmGcTime,
+      metrics.resultSerializationTime,
+      metrics.memoryBytesSpilled,
+      metrics.diskBytesSpilled,
+      metrics.peakExecutionMemory,
+      metrics.inputBytesRead,
+      metrics.inputRecordsRead,
+      metrics.outputBytesWritten,
+      metrics.outputRecordsWritten,
+      metrics.shuffleRemoteBlocksFetched,
+      metrics.shuffleLocalBlocksFetched,
+      metrics.shuffleFetchWaitTime,
+      metrics.shuffleRemoteBytesRead,
+      metrics.shuffleRemoteBytesReadToDisk,
+      metrics.shuffleLocalBytesRead,
+      metrics.shuffleRecordsRead,
+      metrics.shuffleBytesWritten,
+      metrics.shuffleWriteTime,
+      metrics.shuffleRecordsWritten,
+
+      stageId,
+      stageAttemptId)
+  }
+
+  override protected def doUpdate(): Any = {
+    buildUpdate()
+  }
+
+  def updateAndGet(kvstore: KVStore, now: Long): TaskDataWrapper = {
```

Comment:
why do we need it? Since the caller side doesn't care about the return value, we can just update the `lastWriteTime` in `doUpdate`, and then call `doUpdate` instead.